pkg/cmd/experimental/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,247 @@
+
+			kcmdutil.CheckErr(err)
+			if failed {
+				os.Exit(255)
+			}
+
+		},
+	}
+	cmd.SetOutput(out) // for output re: usage / help
+
+	o.ClientFlags = flag.NewFlagSet("client", flag.ContinueOnError) // hide the extensive set of client flags
+	o.Factory = osclientcmd.New(o.ClientFlags)                      // that would otherwise be added to this command
+	cmd.Flags().AddFlag(o.ClientFlags.Lookup(config.OpenShiftConfigFlagName))
+	cmd.Flags().AddFlag(o.ClientFlags.Lookup("context")) // TODO: find k8s constant
+	cmd.Flags().StringVar(&o.ClientClusterContext, options.FlagClusterContextName, "", "client context to use for cluster administrator")
+	cmd.Flags().StringVar(&o.MasterConfigLocation, options.FlagMasterConfigName, "", "path to master config file (implies --host)")
+	cmd.Flags().StringVar(&o.NodeConfigLocation, options.FlagNodeConfigName, "", "path to node config file (implies --host)")
+	cmd.Flags().BoolVar(&o.IsHost, options.FlagIsHostName, false, "look for systemd and journald units even without master/node config")
+	flagtypes.GLog(cmd.Flags())
+	options.BindLoggerOptionFlags(cmd.Flags(), o.LogOptions, options.RecommendedLoggerOptionFlags())
+	options.BindDiagnosticFlag(cmd.Flags(), &o.RequestedDiagnostics, options.NewRecommendedDiagnosticFlag())
+
+	return cmd
+}
+
+func (o *DiagnosticsOptions) Complete() error {
+	var err error
+	o.Logger, err = o.LogOptions.NewLogger()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o DiagnosticsOptions) RunDiagnostics() (bool, error, int, int) {
+	failed := false
+	warnings := []error{}
+	errors := []error{}
+	diagnostics := map[string][]types.Diagnostic{}
+	AvailableDiagnostics := util.NewStringSet()
+	AvailableDiagnostics.Insert(AvailableClientDiagnostics.List()...)
+	AvailableDiagnostics.Insert(AvailableClusterDiagnostics.List()...)
+	AvailableDiagnostics.Insert(AvailableHostDiagnostics.List()...)
+	if len(o.RequestedDiagnostics) == 0 {
+		o.RequestedDiagnostics = AvailableDiagnostics.List()
+	} else if common := intersection(util.NewStringSet(o.RequestedDiagnostics...), AvailableDiagnostics); len(common) == 0 {
+		o.Logger.Errort("emptyReqDiag", "None of the requested diagnostics are available:\n  {{.requested}}\nPlease try from the following:\n  {{.available}}",
+			log.Hash{"requested": o.RequestedDiagnostics, "available": AvailableDiagnostics.List()})
+		return false, fmt.Errorf("No requested diagnostics available"), 0, 1
+	} else if len(common) < len(o.RequestedDiagnostics) {
+		errors = append(errors, fmt.Errorf("Not all requested diagnostics are available"))
+		o.Logger.Errort("notAllReqDiag", `
+Of the requested diagnostics:
+    {{.requested}}
+only these are available:
+    {{.common}}
+The list of all possible is:
+    {{.available}}
+		`, log.Hash{"requested": o.RequestedDiagnostics, "common": common.List(), "available": AvailableDiagnostics.List()})
+	}
+
+	func() { // don't trust discovery/build of diagnostics; wrap panic nicely in case of developer error
```

Comment:
> But that's client code. Allowing your diagnostics to crash for an end user with a bare stack trace and no context destroys credibility. Don't forget this will be the first line of support.

I'm unsold.  How about we get the non-contentious parts in first and add this back after the main body of work is merged?